momaland/learning/continuous_ppo/momappo_crazyrl.py

```diff
@@ -0,0 +1,541 @@
+    obs: jnp.ndarray  # shape is (num_agents, obs_dim)
+    global_obs: jnp.ndarray  # shape is (global_obs_dim)
+    info: jnp.ndarray
+
+
+class Buffer:
+    """A numpy buffer to accumulate the samples, normally faster than jax based because mutable."""
+
+    def __init__(self, batch_size: int, joint_actions_shape, obs_shape, global_obs_shape, num_agents):
+        """Buffer initialization to keep track of data between episodes."""
+        self.batch_size = batch_size
+        self.joint_actions = np.zeros((batch_size, *joint_actions_shape))
+        self.obs = np.zeros((batch_size, *obs_shape))
+        self.global_obs = np.zeros((batch_size, *global_obs_shape))
+        self.value = np.zeros(batch_size)
+        self.reward = np.zeros(batch_size)
+        self.log_prob = np.zeros((batch_size, num_agents))
+        self.terminated = np.zeros(batch_size)
+        self.info = np.zeros(batch_size)  # TODO
+        self.idx = 0
+
+    def add(
+        self,
+        terminated: bool,
+        joint_actions: np.ndarray,
+        obs: np.ndarray,
+        global_obs: np.ndarray,
+        value: float,
+        reward: float,
+        log_prob: np.ndarray,
+        info: dict,
+    ):
+        """Appending new data to the buffer."""
+        self.terminated[self.idx] = terminated
+        self.joint_actions[self.idx] = joint_actions
+        self.obs[self.idx] = obs
+        self.global_obs[self.idx] = global_obs
+        self.value[self.idx] = value
+        self.reward[self.idx] = reward
+        self.log_prob[self.idx] = log_prob
+        # TODO self.info[self.idx] = info
+        self.idx += 1
+
+    def flush(self):
+        """Resetting the idx."""
+        self.idx = 0
+
+    def to_transition(self):
+        """Cast to type Transition."""
+        return Transition(
+            terminated=jnp.array(self.terminated),
+            joint_actions=jnp.array(self.joint_actions),
+            obs=jnp.array(self.obs),
+            global_obs=jnp.array(self.global_obs),
+            value=jnp.array(self.value),
+            reward=jnp.array(self.reward),
+            log_prob=jnp.array(self.log_prob),
+            info=jnp.array(self.info),
+        )
+
+
+def train(args, weights: np.ndarray, key: chex.PRNGKey):
+    """Main training loop for MOMAPPO with SO collapse."""
+    num_updates = int(args.total_timesteps // args.num_steps)
+    minibatch_size = int(args.num_steps // args.num_minibatches)
+
+    def linear_schedule(count):
+        frac = 1.0 - (count // (args.num_minibatches * args.update_epochs)) / num_updates
+        return args.lr * frac
+
+    env: ParallelEnv = Catch.parallel_env()
+    env = clip_actions_v0(env)
+    env = normalize_obs_v0(env, env_min=-1.0, env_max=1.0)
+    env = agent_indicator_v0(env)
+    for agent in env.possible_agents:
+        for idx in range(env.unwrapped.reward_space(agent).shape[0]):  # reward space still not accessible? @ffelten
+            env = NormalizeReward(env, agent, idx)
```

Comment:
@ffelten the supersuit wrappers (`clip_actions_v0`, `normalize_obs_v0`, and `agent_indicator_v0`) are returning an env of type `pettingzoo.utils.conversions.aec_to_parallel_wrapper`, which does not have a `__get_attr__()` function. That is why I have to unwrap the env here.